### PR TITLE
feat(aio): make TOC floating on wide screens

### PR DESCRIPTION
**Pull request by gkalpak**
POC - alternative to #16530. 
Uses the same element for both floating and embeded TOC and positions via CSS. Does not rely on hard-coded values. Still jumps a bit when you scroll fast 😕

### CODE REVIEW

**Comment by petebacondarwin** [comment]
I'm afraid the jank is just too much for me to approve this :-(

**Comment by gkalpak** [comment]
What if we defer the re-positioning of the TOC until scrolling is over and have it animate in place?
I agree that having its top stay fixed would be best, but I can't think of a way to do this without placing the TOC outside of `md-sidenav-content` 😞

**Comment by petebacondarwin** [comment]
Still doesn't work for me. :-(
I wonder if we could hack the A-M container somehow to get rid of this transform that is stopping us from doing this simply as `fixed` position?

**Comment by gkalpak** [comment]
We can easily remove the `transform`, but I expect they put it there for a reason.
If nothing else, it may negatively affect other animations.

**Comment by gkalpak** [comment]
For easy reference: 
Initial jumpy thing: [link][1]
Deferred repositioning: [link][2]
Fixed on viewport: [link][3]

[1]: https://pr16536-12038740e83f1207103cc469a2e021ff4ee76647.ngbuilds.io/
[2]: https://pr16536-469737b9ad1681e918da40bcb40c274be4ee0fce.ngbuilds.io/
[3]: https://pr16536-92d46c5ab936b42d0934c5151d187a1b2c267e75.ngbuilds.io/

File: aio/src/app/layout/doc-viewer/doc-viewer.component.ts [hunk 1]
```diff
@@ -102,7 +102,7 @@ export class DocViewerComponent implements DoCheck, OnDestroy {
       title = titleEl.innerText.trim();
       if (!/(no-toc|notoc)/i.test(titleEl.className)) {
         this.tocService.genToc(this.hostElement, docId);
-        titleEl.insertAdjacentHTML('afterend', '<aio-toc class="embedded"></aio-toc>');
+        titleEl.insertAdjacentHTML('afterend', '<aio-toc track-top="footer"></aio-toc>');
```
Comment (petebacondarwin) [inline aio/src/app/layout/doc-viewer/doc-viewer.component.ts]: I don't think we want this change. This line is for the embedded ToC that appears within the content when the window is narrow.

File: aio/src/app/embedded/toc/toc.component.ts [hunk 2]
```diff
@@ -41,8 +46,8 @@ export class TocComponent implements OnInit {
     this.isClosed = !this.isClosed;
   }
 
-  // @HostListener('window:scroll')
-  // private onScroll() {
-  //   this.scrollTop = window && window.pageYOffset;
-  // }
+  @HostListener('window:scroll')
+  private onScroll() {
+    this.bottom = this.elemTrackedForTop && `calc(100vh - ${this.elemTrackedForTop.getBoundingClientRect().top}px)`;
```
Comment (petebacondarwin) [inline aio/src/app/embedded/toc/toc.component.ts]: This looks much nicer!

File: aio/src/app/embedded/toc/toc.component.ts [hunk 3]
```diff
@@ -1,4 +1,4 @@
-import { Component,  ElementRef, HostBinding, HostListener, OnInit } from '@angular/core';
+import { Component,  ElementRef, HostBinding, HostListener, Input, OnInit } from '@angular/core';
```
Comment (petebacondarwin) [inline aio/src/app/embedded/toc/toc.component.ts]: Input is not used

**Review by petebacondarwin — CHANGES_REQUESTED** [review]

File: packages/empty.ts [hunk 4]
```diff
@@ -1,12 +0,0 @@
...
- * Copyright Google Inc. All Rights Reserved.
- *
- * Use of this source code is governed by an MIT-style license that can be
- * found in the LICENSE file at https://angular.io/license
- */
-
-// This file intentionally left blank. It is used to load nothing in some cases.
-// Such as parse5/index is redirected here instead of loading into browser.
-
-export let __empty__: any;
```
Comment (petebacondarwin) [inline packages/empty.ts]: Did you really mean to remove this?

File: aio/src/app/layout/doc-viewer/doc-viewer.component.ts (hunk 1 above)
Comment (gkalpak) [inline aio/src/app/layout/doc-viewer/doc-viewer.component.ts]: This is the only ToC :smiley:

**Comment by petebacondarwin** [comment]
Let's go with #16530

**Comment by gkalpak** [comment]
😢

**Pull request closed without merging**